protocol/superc20/mirror-factory-limited.md

```diff
@@ -0,0 +1,90 @@
+
+# Proposed Solution
+
+Limiting the scope to `OptimismMintableERC20Tokens` allows using `CREATE2` instead of `CREATE3`, which is more expensive and complex. This is because the variables CREATE2 uses to define the address of the deployed contract provides all the properties we need to maintain the same address accross chains and easily verify whether a mirror was deployed by the official factory or not:
+
+- `Salt`: If we use the `token` address, we can link the `Mirror` to this `token`. This will allow other contracts to know whether a caller is a legitimate `Mirror` or not.
+- `Sender Address`: In this case, it will be the Factory address itself, or the address of the factory's proxy. This works well, as it the factory can be deployed at the same address accross different chains. For chains in the superchain, it will be a predeploy. This predeploy, however, won't have a pretty address. Instead, it has to be deployed at the address where the L1 Factory is deployed so the address is maintained for all chains.
+- `Init Code`: The same proxy's init code must be used accross chains.
+
+The salt doesn't need to be more than the `OptimismMintableERC20Token` address. The `name` and `symbol`, and `decimals` can be fetched with the `token` address alone in the implementation. The factory will be the `Mirror`'s admin, and the proxy will have to either call `setStorage` immediately with the `token` at the time of deployment, or add a `TOKEN_SLOT` and update it on the constructor.
+
+The factory should be upgradable or have setters to allow the modification of the implementations for the current `Mirrors`. The deployment and update of already-deployed `Mirrors` to new implementations should be open to anyone.
+
+We can track versioning through numbers or the addresses of the implementations used.
+
+This design allows the `StandardBridge` or any contract that needs to check the legitimacy of a `Mirror` to call `MIRROR_FACTORY.getMirrorAddress()` and validate whether the calling user is a mirror that corresponds to a given `token`. As an illustrative example of a potential function of the `StandardBridge`:
+
+```solidity
+function mirrorMint(address _token, uint256 _amount) external {
+	address _mirror = MIRROR_FACTORY.getMirrorAddress(_token);
+	if (_mirror != msg.sender) revert();
+	///...mint logic
+}
+
+```
+
+Example code:
+
+```solidity
+contract MirrorFactory {
+	string public constant MIRROR_VERSION = "1.0.0";
+	Mirror public constant MIRROR_IMPL = <PLACEHOLDER_ADDRESS>;
+
+	function deployMirror(address _token) external returns (address) {
+		require(_isOptimismMintableERC20(_token), "not OptimismMintableERC20Token");
+
+		// Note: There's no _token slot in this proxy
+		Proxy _mirror = address(new L2ChugSplashProxy{salt: keccak256(abi.encode(_token))}(address(this), _token));
+
+		// Note: This function does not exist on L2ChugSplashProxy.
+		_mirror.setImplementation(MIRROR_IMPL);
+	}
+
+	function upgradeMirror(address _mirror) external {
```

Comment:
We should remove this function and think about a 1967 beacon based design for simplicity.

We don't yet have a predeploy system for 1967 proxies, but thinking of making a predeploy that looks like this:

```solidity
contract Mirror1967Beacon {
  function implementation() {
      return address(0x..).code;
  }
}
```

The above contract would be a predeploy and it would hardcode in an address that it returns code of. As part of a hardfork, we deploy smart contracts. We would deploy the implementation and we would be able to know the address ahead of time, see [here](https://github.com/ethereum-optimism/specs/blob/main/specs/fjord/derivation.md#gaspriceoracle-deployment) for previous work on how we can know the address ahead of time. We would be able to install the bytecode into the genesis for new chains

Reply:
What this would allow us to do is to hardfork to update the `Mirror1967Beacon` predeploy and modify the hardcoded address that it returns

Reply:
Perfect, this makes a lot of sense. Thanks for the suggestion, will update this.

Reply:
Another thought, we could do something like this:

```solidity
contract MirrorFactory {
  address implementation;
  constructor() {
    implementation = new MirrorSuperchainERC20();
  }

  function deploy() {
    proxy = new Proxy({ beacon: address(this) });
  }
}
```

Reply:
Agreed with @tynes, my read of this design doc makes it seem like we want to use a beacon pattern here